### PR TITLE
Introduce auto-disposable tasks for ScheduledExecutor

### PR DESCRIPTION
**Pull request by peterjot**
Issue: https://github.com/hazelcast/hazelcast/issues/16994

Hazelcast client protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/330

### CODE REVIEW

**Comment by tkountis** [comment]
Thanks for the PR @peterjot - I will have a look at it tomorrow.

**Comment by tkountis** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1512/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1512/Hazelcast-pr-builder-1512.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CodecHashTest.testCodecHashCodes:53->testCodecsInternal:78 ScheduledExecutorSubmitToMemberCodec.java modified by hand. Instead, use hazelcast-client-protocol to generate this codec. Cause: Hash code does not match with the file content. expected:<[d861dd5ff870e367f1d75507ffd1f511]> but was:<[3a3d6458aba40a886baf0b1eb0d28c42]>
[INFO] 
[ERROR] Tests run: 31127, Failures: 1, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by peterjot** [comment]
@tkountis just regenerated that ^^

**Comment by tkountis** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java [hunk 1]
```diff
@@ -382,12 +394,19 @@ private int getTaskOrKeyPartitionId(Runnable task, Object key) {
...
-        if (command instanceof NamedTask) {
-            name = ((NamedTask) command).getName();
-        }
+        String taskName = getNamedTaskName(command);
+        return taskName != null ? taskName : UuidUtil.newUnsecureUuidString();
+    }
 
-        return name != null ? name : UuidUtil.newUnsecureUuidString();
+    private String getNamedTaskName(Object command) {
+        if (command instanceof AbstractTaskDecorator) {
```
Comment (tkountis) [inline hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java:402]: I like what you did here, but unfortunately the `NamedTask` marker is public, meaning anyone can implement it directly without the use of our helper. In this case, the `AbstractTaskDecorator` inheritance expectation is not going to be valid.

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java [hunk 2]
```diff
@@ -433,4 +452,11 @@ private String extractNameOrGenerateOne(Object command) {
             throw rethrow(e);
         }
     }
+
+    private boolean isAutoDisposable(Object command) {
+        if (command instanceof AbstractTaskDecorator) {
```
Comment (tkountis) [inline hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java:457]: Same here.

File: hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceProxy.java [hunk 3]
```diff
@@ -493,11 +503,10 @@ private String extractNameOrGenerateOne(Object command) {
         return createFutureProxy(uuid, taskName);
     }
 
-    @SuppressWarnings("unchecked")
     private <T> T initializeManagedContext(Object object) {
         ManagedContext context = getNodeEngine().getSerializationService().getManagedContext();
-        if (object instanceof NamedTaskDecorator) {
-            ((NamedTaskDecorator) object).initializeContext(context);
+        if (object instanceof AbstractTaskDecorator) {
```
Comment (tkountis) [inline hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceProxy.java:508]: Same concerns as client proxy

File: hazelcast/src/test/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceBasicTest.java [hunk 4]
```diff
@@ -874,6 +904,49 @@ public void schedule_thenCancelAndGet() throws Exception {
...
+    @Test
+    public void schedule_whenAutoDisposable_thenGet() throws Exception {
+        HazelcastInstance[] instances = createClusterWithCount(2);
+        IScheduledExecutorService executorService = getScheduledExecutor(instances, "s");
+
+        IScheduledFuture<Double> future = executorService.schedule(autoDisposable(new PlainCallableTask()), 1, SECONDS);
+
+        expected.expect(ExecutionException.class);
+        expected.expectCause(isA(StaleTaskException.class));
+        Thread.sleep(2000);
```
Comment (tkountis) [inline hazelcast/src/test/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceBasicTest.java:916]: `Thread.sleep` as you already mentioned is prone to cause flakiness. See `assertTrueEventually` which works with the maximum test waiting time.

**Comment by tkountis** [comment]
@peterjot great work, some comments to consider, the rest looks good. 
I will have another look at it tomorrow to consider edge cases, but it looks ok. :+1:

File: hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java (hunk 2 above)
Comment (peterjot) [reply hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientScheduledExecutorProxy.java:457]: I think in that case this is handled at line 460

File: hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceProxy.java (hunk 3 above)
Comment (peterjot) [reply hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorServiceProxy.java:508]: It seems to me that an object implementing NamedTask is handled by the `else` block. As it worked before.

**Comment by peterjot** [comment]
Thank you very much @tkountis for the review! I have fixed these tests and handled NamedTask implementations

**Comment by tkountis** [comment]
LGTM 
One more thing, could you please add a test for the NamedTask on the scheduler, ideally I want to see that you can retrieve (see `getScheduledFuture(handler)`) a scheduled task by its name whether its scheduled with a marker interface or with the wrapper class you created.

**Comment by tkountis** [comment]
run-lab-run

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1594/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1594/Hazelcast-pr-builder-1594.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   MemberCompatibilityNullTest_2_1.test_ScheduledExecutorSubmitToMemberCodec_decodeRequest:6003
[ERROR]   MemberCompatibilityNullTest_2_1.test_ScheduledExecutorSubmitToPartitionCodec_decodeRequest:5979
[ERROR]   MemberCompatibilityTest_2_1.test_ScheduledExecutorSubmitToMemberCodec_decodeRequest:6003
[ERROR]   MemberCompatibilityTest_2_1.test_ScheduledExecutorSubmitToPartitionCodec_decodeRequest:5979
[INFO] 
[ERROR] Tests run: 31167, Failures: 4, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by tkountis** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
Thank you for the contribution @peterjot ! cc @Holmistr

**Comment by yquest** [comment]
Hi guys I don't know if you've noticed but the autodisposable doesn't work for Runnables in version 4.2.2, only works for Callables

**Comment by mmedenjak** [comment]
Hi @yquest ! Do you have a reproducer? Can you open an issue with it?

**Comment by yquest** [comment]
> Hi @yquest ! Do you have a reproducer? Can you open an issue with it?

Ok, I did that, I create the issue 19622

**Pull request merged**